### ISSUE TITLE
Opinions on JavaScript transpiler and preprocessor usages

### ISSUE THREAD

**Issue by rockacola** [CityOfZion/neo-js#102]
I would like to yield interests and opinions on pushing the project a step closer to defacto-standard, at a cost of increasing learning curve. There are potential we can result in even better coding practises and decrease environment requirements (particularly browsers).

**Comment by rockacola** [comment]
My vote goes to: TypeScript with NodeJS as build process.

Reasons:
* Stricter variable/method declarations, coupled with tigher linting rules.
* If we are to use a transpiler to increase ES6 support on browsers, then you may find TypeScript transpiler to be more reliable than Babel, especially when dealing with edge cases.
* I don't see a need to introduce advanced preprocessor like webpack or gulp for this project, therefore by staying with NodeJS we can reduce the learning curve as well as remove 1 potential point of failure.

**Comment by ClarenceL** [comment]
Personally not a fan of TypeScript, just feels overburdening, Python is even more loosely typed and neo-python is coming along just fine.

Though I guess there is a lot of cooperation with Microsoft so it may make sense to go that route, would like to hear other opinions on Typescript.

**Comment by rockacola** [comment]
Not disagreeing with you, but to add support arguments for pro TypeScript:

* This JavaScript project encourges documentation through code comments. TypeScript further that by enforcing correct input argument types and output type at transpile level.
* My personal favourite benefit of TypeScript is its `strictNullChecks: true` capability: explicit declarable whether a variable can contain `null` or not. This can help reduce a lot of unnecessary (but reasonable) verifications in an open source project where a contributor may not be well awared full extend of all internal methods used.

**Comment by notatestuser** [comment]
I agree that TypeScript is overburdening for some existing projects but the extent of this depends on many factors and I'm not suggesting anything in particular for neo-js.

However, I've had some success with [flow](https://github.com/facebook/flow) with [flow-jsdoc](https://github.com/Kegsay/flow-jsdoc) in my (unpushed) neon-js experiments. This is an alternative to TypeScript that would:
- Use existing jsdoc comments and not require new explicit type annotations in code. jsdoc is used in these projects so the type annotations are already there and should not have to be duplicated.
- Run like a linter in the console or in an editor. Flow is essentially a linter with type inference through data flow analysis magic.
- Require minimal code changes, adding the most value with little effort.

To address the requirements @rockacola has stated already:
- Tighter linting - flow checks types on declarations and throws errors if usages do not match up.
- Babel for browser support still possible - no code changes required as types come from jsdoc.
- "strictNullChecks": flow will throw errors about type incompatibilities of null vs. type.

**Comment by rockacola** [comment]
Thanks again @ClarenceL and @notatestuser for your input. I can lean toward Flow, for it been the smallest learning curve.

For version `0.x`, we won't be consider about browser support and this project will purely design for Node.js usage. We'll revisit this topic again when we reached `1.x`. Hence there won't be a need Babel for now.

**Comment by rockacola** [comment]
@notatestuser would you like to look into Flow usage integration for the project as part of `0.7.0`?

**Comment by notatestuser** [comment]
Sure, I would be happy to do that. Keen on hearing @ClarenceL’s thoughts on this too. Flow coverage is on a per file basis so we could spread the workload after initial setup is there.

**Comment by rockacola** [comment]
If our own is to prevent the need for transpilers, then my own known option is to adapt `flow-jsdoc`.

**Comment by notatestuser** [comment]
I have done some work on this in my branch. Should have a PR soon.

**Comment by rockacola** [comment]
Please be informed a major structural updates as of PR #132

I don't mind to either take your suggestions and refit into the new structure, or allow you to resync with the change once it is merged.

**Comment by rockacola** [comment]
Was not exactly convinced with `flow-jsdoc` base on my attempt for adoptation. More feedback on `flow-jsdoc` or other transpilers/preprocessors are welcome.

Postpone this case from `0.7.0` milestone.

**Comment by rockacola** [comment]
Not reaching to any conclusion, will stick with existing ES6 + JSDoc process for now.
